File: cpp/69_totient_maximum.cpp
#include <iostream>
#include <vector>

int totient_maximum(){
    std::vector<double> totients(1'000'001, 0);
    for (int i = 0; i < totients.size(); i++) totients[i] = i;
    std::vector<bool> sieved(1'000'001, false);
    for (int i = 2; i < sieved.size(); i++){
        if (!sieved[i]){
            // totient(n) = n - 1 if n is prime
            totients[i]--;
            for (int j = 2 * i; j < sieved.size(); j += i){
                sieved[j] = true;
                // totient(n) = n * P(p - 1/p)
                // n * product of (p - 1/p) for all distinct prime factors p of n
                
                // find prime factors p of n
                // multiply current stored totient value of n by (p - 1) / p
                totients[j] *= (i - 1.0) / i;
            }
        }
    }
    int res = 0;
    double maxVal = 0;
    for (int i = 2; i < totients.size(); i++){
        double val = i / totients[i];
        if (val > maxVal){
            res = i;
            maxVal = val;
        }
    }
    return res;
}

int main(){
    std::cout << totient_maximum() << "\n";
}

File: c#/_69_TotientMaximum.cs
using System;

public class _69_TotientMaximum{
	public static int Euler69(){
		int n = 1_000_000, res = 0;
		double maxRatio = 1;
		bool[] sieved = new bool[n + 1];
		double[] totients = new double[n + 1];
		for (int i = 0; i <= n; i++){
			totients[i] = i;
		}
		for (int i = 2; i <= n; i++){
			if (!sieved[i]){
				totients[i]--;
				for (int j = 2 * i; j <= n; j += i){
					sieved[j] = true;
					totients[j] *= (i - 1.0) / i;
				}
			}
			double ratio = i / totients[i];
			if (ratio > maxRatio){
				res = i;
				maxRatio = ratio;
			}
		}
		return res;
	}
	
	public static void Main(){
		Console.WriteLine(Euler69());
	}
}
